### PR TITLE
drawio-image/figure directives to replace drawio

### PR DESCRIPTION
**Pull request by brechtm**
- inherits from the image/figure directives so that draw.io images
  behave the same as images/figures
- this means that all builders will be able to handle draw.io images
  without any custom code

Refer to the README for an overview of the changes compared to the old
drawio directive. See also #28.

Diagrams are now also exported to the .drawio directory beside the
document. Each diagram is exported to a separate subdirectory in this directory named by the hash of the diagram filename and export
options. The base name of the exported image file is the same as that
of the draw.io file. This has these advantages:
- we don't interfere with the image/figure directive copying of image
  files to the output directory
- a diagram needs to be exported only once when two builders use the
  same export format
- more descriptive image filenames for the HTML builder

### CODE REVIEW

**Comment by brechtm** [comment]
Two tests fail when the whole test suit is run at once (like on CircleCI) for some reason. I'll investigate that.

File: sphinxcontrib/drawio.py
Comment (Modelmat) [inline sphinxcontrib/drawio.py:272]: I'd prefer if this was a normal dictionary expression, i.e.:
```suggestion
DEFAULT_BUILDER_EXPORT_FORMAT = {
	"html": "svg",
	"latex": "pdf",
	"rinoh": "pdf",
}
```

File: sphinxcontrib/drawio.py
Comment (Modelmat) [inline sphinxcontrib/drawio.py:174]: I'd prefer this not to just be `folder-1.png`. If possible:
* use the filename of the draw.io file; as the tests have it'd be `box`.
* if the options are changed from the config **defaults**, append them in alphabetical order.

e.g. for a file named box with the zoom changed from 100 to 150:
`box-scale150.png`

File: README.md
Comment (Modelmat) [inline README.md:7]: Please add a note that these will be removed in version 1.0 [I think this makes sense?]

**Review by Modelmat — CHANGES_REQUESTED** [review]
A few requests.

If possible, could you make git recognise the **move** of `drawio.py` to `old_drawio` as a move, and the creation of a new `drawio.py`? Should be able to use `git mv` or somewhat (might need a rebase, and _possibly_ multiple comments but I don't think so).

I'd also prefer if the old file is `old_drawio` not `olddrawio`.

**Comment by brechtm** [comment]
Thanks for the quick review!

> If possible, could you make git recognise the **move** of `drawio.py` to `old_drawio` as a move, and the creation of a new `drawio.py`? Should be able to use `git mv` or somewhat (might need a rebase, and _possibly_ multiple comments but I don't think so).

I'm afraid not. `git mv` doesn't do anything special; no "move" or "copy" information is stored in commits, AFAIK. Git identifies file copies after the fact, for example when you do `git log -p`.

> I'd also prefer if the old file is `old_drawio` not `olddrawio`.

I'll change that. I also considered creating a `drawio` package with `__init__.py` (new code) and `deprecated.py` (old code). Let me know if you prefer that approach.

File: requirements.txt
Comment (brechtm) [reply requirements.txt:1]: As it is now, we're not testing against the current Sphinx version at all. Ideally, we'd test against a couple of Sphinx versions. Since implementing that is a sizeable project on its own, I figured we should preferably test against the current version. We can drop the `==3.2.1` to always use the latest version.

File: sphinxcontrib/drawio.py
Comment (brechtm) [reply sphinxcontrib/drawio.py:68]: This class is no longer needed. I'm removing the _get_filename_ function I introduced since it's only called in a single place now.

File: sphinxcontrib/drawio.py
Comment (brechtm) [reply sphinxcontrib/drawio.py:88]: Changed. Old habits die hard :-)
We should look into using [black](https://github.com/psf/black).

File: sphinxcontrib/drawio.py
Comment (brechtm) [reply sphinxcontrib/drawio.py:195]: Done.

File: sphinxcontrib/drawio.py
Comment (brechtm) [reply sphinxcontrib/drawio.py:140]: Done.

File: sphinxcontrib/drawio.py
Comment (brechtm) [reply sphinxcontrib/drawio.py:213]: Correct. As far as the builders are concerned there are no new nodes.

File: sphinxcontrib/drawio.py
Comment (brechtm) [reply sphinxcontrib/drawio.py:272]: Done.

File: sphinxcontrib/drawio.py
Comment (brechtm) [reply sphinxcontrib/drawio.py:174]: I'm not sure I'm following you. `box.drawio` will be exported to `.drawio/<hash A>/box.png`. The HTML builder will copy that file to `_build/_html/_images/box.png`. If`box.drawio` is referenced in another `drawio-image` directive with different export options, it will be exported to ``.drawio/<hash B>/box.png`. The HTML builder copies that file to `_build/_html/_images/box1.png`.

I feel that adding the export options to the filename:
- is not so useful since in most cases you will only reference a diagram once,
- makes image filenames unnecessarily long, and
- outside the scope of this PR

File: README.md
Comment (brechtm) [reply README.md:7]: Done.

File: sphinxcontrib/drawio.py
Comment (brechtm) [reply sphinxcontrib/drawio.py:120]: Replaced with more intelligent code.

File: requirements.txt
Comment (Modelmat) [reply requirements.txt:1]: Doing `sphinx>=2` would install the latest version I'm pretty sure, so it would work.

File: sphinxcontrib/drawio.py
Comment (Modelmat) [reply sphinxcontrib/drawio.py]: Yeah, I'm pretty sure I had it as an issue to add it. Apparently not => #37

File: sphinxcontrib/drawio.py
Comment (Modelmat) [reply sphinxcontrib/drawio.py:161]: Okay then.
This is behaviour seems to be that of the image figure backend, yeah?

File: sphinxcontrib/drawio.py
Comment (Modelmat) [inline sphinxcontrib/drawio.py:107]:
```suggestion
            if isinstance(node, docutils_image):
```

File: sphinxcontrib/drawio.py
Comment (brechtm) [reply sphinxcontrib/drawio.py:161]: Which behavior you are referring to? If you're talking about the copying of the image files, that is how the HTML builder handles image nodes.

Comment (Modelmat) [reply sphinxcontrib/drawio.py:161]: Yeah, that's what I was trying to get at. That's fine then.

**Comment by Modelmat** [comment]
> I'm afraid not. git mv doesn't do anything special; no "move" or "copy" information is stored in commits, AFAIK. Git identifies file copies after the fact, for example when you do git log -p.

Okay.

> I'll change that. I also considered creating a drawio package with __init__.py (new code) and deprecated.py (old code). Let me know if you prefer that approach.

The current one is fine.

---
I'd thought I'd already responded to this, I probably forgot to press send ;)

**Review by Modelmat — APPROVED** [review]

**Comment by Modelmat** [comment]
If you'd also like to rebase your changes onto latest master, that'd be great :)

**Comment by brechtm** [comment]
I rebased and renamed _olddrawio.py_.

Thanks for the push access! How about I commit bugfixes myself but still open a PR for feature additions so they are first reviewed?

**Comment by Modelmat** [comment]
I think it's probably best to keep everything to PRs, however it probably depends on the nature of the fix - a one line change vs refactoring some method would make a difference to whether it's okay (the first would be fine, the latter, probably not).
That being said, if I take 2 weeks to respond to an bugfix PR you can probably just merge it :)

I'll merge this PR when I'm next at a desktop, and will probably tag a new version on pypi.

**Comment by brechtm** [comment]
Just a friendly reminder to merge this and create a new release. We're anxious to start using it :-)

**Comment by Modelmat** [comment]
Sorry, my bad. I'm in the middle of exams right now.

If you'd like to merge & tag a new release go for it.

[Not sure if documented but I tend to use lightweight tags on the latest commit then push both tags & commit at the same time].

**Comment by Modelmat** [comment]
*Just to be clear, I'm saying that you _should_ merge and tag this because I won't get to it for a while.

**Comment by brechtm** [comment]
> Sorry, my bad. I'm in the middle of exams right now.

No problem. Good luck!

Merged, tagged, pushed: 0.0.10 is available on PyPI! We'll start using this now and I'll be the fixing bug we run into. Cheers!

**Pull request merged**
